smartapps/curb/curb-energy-monitor.src/curb-energy-monitor.groovy

```diff
@@ -0,0 +1,620 @@
+            }
+        }
+    } else {
+        return dynamicPage(name: "auth", title: "Login", nextPage: "", uninstall: false) {
+            section() {
+                paragraph("Tap below to log in to the CURB service and authorize SmartThings access")
+                href url: buildRedirectUrl, style: "embedded", required: true, title: "CURB", description: "Click to enter CURB Credentials"
+            }
+        }
+    }
+}
+
+def oauthInitUrl() {
+    atomicState.oauthInitState = UUID.randomUUID().toString()
+    def oauthParams = [
+        response_type: "code",
+        scope: "offline_access",
+        audience: "app.energycurb.com/api",
+        client_id: appSettings.clientId,
+        connection: "Users",
+        state: atomicState.oauthInitState,
+        redirect_uri: callbackUrl
+    ]
+    redirect(location: "${curbLoginUrl}?${toQueryString(oauthParams)}")
+}
+
+def callback() {
+    //log.debug "callback()>> params: $params, params.code ${params.code}"
+    def code = params.code
+    def oauthState = params.state
+    if (oauthState == atomicState.oauthInitState) {
+        def tokenParams = [
+            grant_type: "authorization_code",
+            code: code,
+            client_id: appSettings.clientId,
+            client_secret: appSettings.clientSecret,
+            redirect_uri: callbackUrl
+        ]
+        httpPostJson([uri: curbTokenUrl, body: tokenParams]) {
+            resp ->
+                //log.debug "response contentType: ${resp.contentType}"
+                //log.debug("Got POST response: ${resp.data}")
+
+                atomicState.refreshToken = resp.data.refresh_token
+            atomicState.authToken = resp.data.access_token
+
+            getCurbLocations()
+        }
+        if (atomicState.authToken) {
+            success()
+        } else {
+            fail()
+        }
+    } else {
+        log.error "callback() failed oauthState != atomicState.oauthInitState"
+    }
+}
+
+private removeChildDevices(delete) {
+    delete.each {
+        deleteChildDevice(it.deviceNetworkId)
+    }
+}
+
+def getCurbLocations() {
+    //log.debug("Requesting Curb location info");
+
+    def params = [
+        uri: "http://app.energycurb.com",
+        path: "/api/locations",
+        headers: ["Authorization": "Bearer ${atomicState.authToken}"]
+    ]
+
+    try {
+        httpGet(params) {
+            resp ->
+                def locationNameList = []
+            def locationLookup = []
+            //log.debug(resp.data)
+            resp.data.each {
+                locationNameList.push(it.name)
+                locationLookup.push(it)
+            }
+            atomicState.locationNames = locationNameList
+            atomicState.locationLookup = locationLookup
+
+            //log.debug("Location ID: ${atomicState.location}")
+        }
+    } catch (e) {
+        log.error "something went wrong: $e"
+    }
+}
+
+def updateSelectedLocationId() {
+    atomicState.locationLookup.each {
```

Comment:
since assigning an atomiState.variable immediately will save the whole state object it's not recommended to loop over the atomicState and make change to it inside the loop. If possible consider to use state instead, just make sure that cases that can cause an update to the state variable can be triggered through multiple instances, otherwise consider changing the code so that atomicState is only update once when the method ends.